### PR TITLE
fix(forms): Add float number support for min and max validator

### LINKED ISSUE #42215: MIN and MAX attributes in Angular 12 templated forms not working with decimal values
# Bug Report

### Affected Package

FormsModule
@angular/forms

Commit: 8fb83ea1b524663bd804908308d528c338450270
* feat(forms): introduce min and max validators (#39063)

validators.ts:

Using parseInt rather than parsefloat ?

```
normalizeInput = (input: string): number => parseInt(input, 10);
```

### Is this a regression?

MIN/MAX attributes didn't participate in templated forms validity < v12 so not a regression as such.

### Description

MIN and MAX attributes not honouring decimal values in v12. In the example below, the form becomes invalid with a maximum over 37, e.g. 37.1 makes form invalid when it should be > 37.5

```html
<form #someform="ngForm">
  <input
    [(ngModel)]="model.formContractHours"
    required
    name="ContractHours"
    type="number"
    [min]="0.1"
    [max]="37.5"
    step="0.1"
  />
</form>
```
## Minimal Reproduction

MIN/MAX attributes ignored in V11 templated form
https://stackblitz.com/edit/angular11minmaxdecimals

MIN/MAX attributes validated in V12 in templated form but not respecting decimals.
https://stackblitz.com/edit/angular12minmaxdecimals

### PR DESCRIPTION
**Pull request by iRealNirmal**
Added float number support for min and max validator

Closes #42215

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Currently min and max validators are not supporting float number

Issue Number: #42215 

## What is the new behavior?
Added float number support for min and max validator

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/forms/test/validators_spec.ts [hunk 1]
```diff
@@ -61,6 +61,20 @@ describe('Validators', () => {
       expect(Validators.min(2)(new FormControl('1'))).toEqual({'min': {'min': 2, 'actual': '1'}});
     });
 
+    it('should not error on small float number validation', () => {
```
Comment (AndrewKushnir) [inline packages/forms/test/validators_spec.ts:64]: The tests added in this PR verify the logic of the underlying `Validators.min` and `Validators.max` validator functions, but they are not checking the behavior of the min/max validator directives that was actually changed. These tests are actually helpful and we should keep them (I didn't find existing tests that verify the logic with float numbers) + we should add tests to verify directives logic as well. You can look at the test cases added in the original PR #39063 (see `packages/forms/test/template_integration_spec.ts` and `packages/forms/test/reactive_integration_spec.ts`) and either update them with additional cases or create a couple new ones. Please let me know if you have any questions, will be happy to help. Thank you.

**Review by AndrewKushnir — COMMENTED** [review]
@iRealNirmal thanks for the fix 👍 

The code looks good, I've just left a comment about extra tests that we'd need to add.

File: packages/forms/test/validators_spec.ts (hunk 1 above)
Comment (iRealNirmal) [reply packages/forms/test/validators_spec.ts:64]: @AndrewKushnir  thanks, that was really good input. While writing test case I realize that still input wasn't allowing float number, so I did code change for it as well. Feel free to provide any other inputs as well.
cc : @petebacondarwin

File: packages/forms/test/reactive_integration_spec.ts [hunk 2]
```diff
@@ -2563,6 +2563,34 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
...
+            dispatchEvent(input, 'input');
+            expect(form.value).toEqual({pin: 10.15});
+            expect(form.valid).toBeTruthy();
+            expect(form.controls.pin.errors).toBeNull();
+
+            fixture.componentInstance.max = 10.05;
+            fixture.detectChanges();
+
+            expect(form.valid).toBeFalse();
+            expect(form.controls.pin.errors).toEqual({max: {max: 10.05, actual: 10.15}});
```
Comment (petebacondarwin) [inline packages/forms/test/reactive_integration_spec.ts:2591]: Maybe now set the value back to a number that will be valid to show that the validator will be run when the value changes? Same for the other `min` tests and those in `template_integration.spec.ts`?

**Review by petebacondarwin — APPROVED** [review]
One small suggestion.

File: packages/forms/test/reactive_integration_spec.ts (hunk 2 above)
Comment (iRealNirmal) [reply packages/forms/test/reactive_integration_spec.ts:2591]: I will move before line 2581 so that aspect will be covered.

Comment (iRealNirmal) [reply packages/forms/test/reactive_integration_spec.ts:2591]: yes seems I missed

File: packages/forms/src/directives/validators.ts [hunk 3]
```diff
@@ -580,7 +580,7 @@ export class MinLengthValidator implements Validator, OnChanges {
 
   private _createValidator(): void {
     this._validator = minLengthValidator(
-        typeof this.minlength === 'number' ? this.minlength : parseInt(this.minlength, 10));
+        typeof this.minlength === 'number' ? this.minlength : parseFloat(this.minlength));
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:583]: @iRealNirmal I think we should revert this change and the change below, since they are related to the `minlength` and `maxlength` validators (not `min`/`max` ones), just keep the changes in `normalizeInput` functions.

Comment (iRealNirmal) [reply packages/forms/src/directives/validators.ts:583]: ohh, yes you are correct. Really good catch.  Seems I got mixed. Let me correct  this.

File: packages/forms/src/directives/validators.ts [hunk 4]
```diff
@@ -657,7 +657,7 @@ export class MaxLengthValidator implements Validator, OnChanges {
 
   private _createValidator(): void {
     this._validator = maxLengthValidator(
-        typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength, 10));
+        typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength));
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:660]: @iRealNirmal could you please revert this change (it looks unrelated)?

**Review by AndrewKushnir — COMMENTED** [review]
@iRealNirmal thanks for adding tests and updating the code 👍 

I've just left one final comment and we should be ready for final checks in Google's codebase and adding this PR to the merge queue if everything goes well.

File: packages/forms/src/directives/validators.ts (hunk 4 above)
Comment (iRealNirmal) [reply packages/forms/src/directives/validators.ts]: yes removed, thanks

**Review by AndrewKushnir — APPROVED** [review]
Looks great, thanks @iRealNirmal 👍 

I've started [tests in Goolge's codebase](http://test/OCL:375735791:BASE:375735831:1621963404178:ff20e7b8) (internal-only link) and will keep this thread updated.

**Comment by AndrewKushnir** [comment]
@iRealNirmal FYI presubmits in Google's codebase went well, I'm adding this PR to the merge queue.
